src/main/java/servico/bookstore/Bookstore.java

```diff
@@ -703,66 +705,78 @@ public Stock getStock(int bId) {
+                .findFirst()
+                .ifPresent(line -> clientIds.add(order.getCustomer().getId()));
         }
-        HashMap<Integer, Counter> counters = new HashMap<>();
-        i = ordersByCreation.iterator();
-        while (i.hasNext()) {
-            Order order = i.next();
-            if (clientIds.contains(order.getCustomer().getId())) {
-                order.getLines().forEach((line) -> {
+
+        return clientIds;
+    }
+
+    private Map<Integer, BookstoreBookCounter> countBooksBoughtByClients(Set<Integer> clientIds, Book targetBook) {
+        Map<Integer, BookstoreBookCounter> counters = new HashMap<>();
+
+        // Para cada ordem, se o cliente comprou o livro alvo, conta os outros livros comprados
+        ordersByCreation.stream()
+            .filter(order -> clientIds.contains(order.getCustomer().getId()))
+            .forEach(order -> {
+                order.getLines().forEach(line -> {
                     Book book = line.getBook();
-                    if (targetBook.getId() != book.getId()) {
-                        Counter counter = counters.get(book.getId());
-                        if (counter == null) {
-                            counter = new Counter();
-                            counter.book = book;
-                            counter.count = 0;
-                            counters.put(book.getId(), counter);
-                        }
-                        counter.count += line.getQty();
+                    if (book.getId() != targetBook.getId()) {
+                        counters.computeIfAbsent(book.getId(), id -> new BookstoreBookCounter(book))
+                                .addQuantity(line.getQty());
                     }
                 });
-            }
-        }
-        Counter[] sorted = counters.values().toArray(new Counter[] {});
-        Arrays.sort(sorted, (Counter a, Counter b) -> {
-            if (b.count > a.count) {
-                return 1;
-            }
-            return b.count < a.count ? -1 : 0;
-        });
-        Book[] related = new Book[] { targetBook, targetBook, targetBook,
-                targetBook, targetBook };
-        for (j = 0; j < 5 && j < sorted.length; j++) {
-            related[j] = sorted[j].book;
+            });
+
+        return counters;
+    }
+
+    private Book[] getTopFiveRelatedBooks(Map<Integer, BookstoreBookCounter> bookCounters, Book targetBook) {
+        // Ordena os livros mais vendidos e pega os cinco primeiros
+        List<BookstoreBookCounter> sortedCounters = bookCounters.values().stream()
+            .sorted(Comparator.comparingInt(BookstoreBookCounter::getCount).reversed())
+            .collect(Collectors.toList());
+
+        // Preenche com o livro alvo, caso existam menos de 5 livros relacionados
+        Book[] relatedBooks = new Book[5];
+        for (int i = 0; i < Math.min(5, sortedCounters.size()); i++) {
+            relatedBooks[i] = sortedCounters.get(i).getBook();

```

Comment:
```suggestion
       int top = 5;
        Book[] relatedBooks = new Book[top];
        for (int i = 0; i < Math.min(top, sortedFrequencies.size()); i++) {
            relatedBooks[i] = sortedFrequencies.get(i).getBook();
```
remvi o comentario e criei uma variavel para não ficar o valor hardcoded, alem disso se quisermos transformar esse cara num  paramentro do metodo no futuro, ja fica na mao tbm